server/app/services/program/ProgramService.java

```diff
@@ -1305,6 +1305,34 @@ public ProgramDefinition removeBlockPredicate(long programId, long blockDefiniti
     }
   }
 
+  /**
+   * Update the eligibility message for a block.
+   *
+   * @param programId the ID of the program to update
+   * @param blockDefinitionId the ID of the block to update
+   * @param message the custom eligibility message to add to the block
+   * @return the updated block
+   * @throws ProgramNotFoundException when programId does not correspond to a real Program.
+   * @throws ProgramBlockDefinitionNotFoundException when blockDefinitionId does not correspond to a
+   *     real Block.
+   */
+  public ProgramDefinition setBlockEligibilityMessage(
+      long programId, long blockDefinitionId, Optional<LocalizedStrings> message)
+      throws ProgramNotFoundException, ProgramBlockDefinitionNotFoundException {
+    ProgramDefinition programDefinition = getFullProgramDefinition(programId);
+
+    BlockDefinition blockDefinition =
+        programDefinition.getBlockDefinition(blockDefinitionId).toBuilder()
+            .setLocalizedEligibilityMessage(message)
+            .build();
+    try {
+      return updateProgramDefinitionWithBlockDefinition(programDefinition, blockDefinition);
```

Comment:
@gwendolyngoetz could something like this be the source of the null questions?

here we're getting a program definition, potentially from the cache but not necessarily, and then modifying it and re-saving it, and there's a potential for a race condition since it's not all in a transaction. Should we wrap this in a transaction?

Reply:
Hmm, yeah maybe! It should definitely get rolled back if it fails if it isn't already.

Reply:
After offline discussion, we're going to do the transaction after the other parts of eligibility message are done. To track this I filed issue #9277.